File: laba4/ex4.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <sstream>
#include <set>

using namespace std;

vector<int> borda_winners(const vector<vector<int>>& prefs, int n) {
    vector<int> scores(n, 0);
    for (const auto& ranking : prefs) {
        for (int pos = 0; pos < n; ++pos) {
            scores[ranking[pos]] += (n - 1 - pos);
        }
    }
    
    int max_score = *max_element(scores.begin(), scores.end());
    vector<int> winners;
    for (int i = 0; i < n; ++i) {
        if (scores[i] == max_score) winners.push_back(i);
    }
    return winners;
}

bool is_condorcet_winner(int c, const vector<vector<int>>& prefs, int n, int k) {
    for (int other = 0; other < n; ++other) {
        if (other == c) continue;
        int wins = 0;
        for (const auto& ranking : prefs) {
            auto c_pos = find(ranking.begin(), ranking.end(), c);
            auto other_pos = find(ranking.begin(), ranking.end(), other);
            if (c_pos < other_pos) wins++;
        }
        if (wins <= k / 2) return false;
    }
    return true;
}

int condorcet_winner(const vector<vector<int>>& prefs, int n, int k) {
    for (int c = 0; c < n; ++c) {
        if (is_condorcet_winner(c, prefs, n, k)) return c;
    }
    return -1;
}

int main() {
    int n, k;
    cout << "Введите количество кандидатов (n): ";
    cin >> n;
    cout << "Введите количество избирателей (k): ";
    cin >> k;
    cin.ignore();

    // Ввод имён кандидатов
    vector<string> candidates(n);
    unordered_map<string, int> candidate_to_idx;
    cout << "Введите имена кандидатов (как есть, без нормализации):\n";
    for (int i = 0; i < n; ++i) {
        cout << "Кандидат " << i+1 << ": ";
        getline(cin, candidates[i]);
        if (candidate_to_idx.count(candidates[i])) {
            cerr << "Ошибка: Кандидат '" << candidates[i] << "' уже существует!";
            return 1;
        }
        candidate_to_idx[candidates[i]] = i;
    }

    // Ввод предпочтений
    vector<vector<int>> prefs(k, vector<int>(n));
    cout << "\nВведите предпочтения (" << n << " имен через пробел):\n";
    for (int i = 0; i < k; ++i) {
        while (true) {
            cout << "Избиратель " << i+1 << ": ";
            string line;
            getline(cin, line);
            stringstream ss(line);
            vector<int> votes;
            string name;
            bool error = false;

            while (ss >> name) {
                if (!candidate_to_idx.count(name)) {
                    cout << "Ошибка: Неизвестный кандидат '" << name << "'\n";
                    error = true;
                    break;
                }
                votes.push_back(candidate_to_idx[name]);
            }

            if (error) continue;
            if (votes.size() != n) {
                cout << "Ошибка: Нужно " << n << " кандидатов!\n";
                continue;
            }

            set<int> unique(votes.begin(), votes.end());
            if (unique.size() != n) {
                cout << "Ошибка: Повторы в списке!\n";
                continue;
            }

            prefs[i] = votes;
            break;
        }
    }

    // Результаты
    vector<int> borda = borda_winners(prefs, n);
    int condorcet = condorcet_winner(prefs, n, k);

    cout << "\nРезультаты:\nМетод Борда: ";
    if (borda.size() == 1) cout << candidates[borda[0]];
    else {
        cout << "Ничья между: ";
        for (size_t i = 0; i < borda.size(); ++i) {
            cout << "'" << candidates[borda[i]] << "'";
            if (i != borda.size()-1) cout << ", ";
        }
    }

    cout << "\nМетод Кондорсе: ";
    if (condorcet != -1) cout << candidates[condorcet];
    else cout << "Нет победителя";

    if (condorcet != -1 && find(borda.begin(), borda.end(), condorcet) == borda.end()) {
        cout << "\n\n⚠️ Методы дали разных победителей!";
    }

    return 0;
}

File: laba4/ex4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

// Borda winners calculation
func bordaWinners(prefs [][]int, n int) []int {
	scores := make([]int, n)
	for _, ranking := range prefs {
		for pos, candidate := range ranking {
			scores[candidate] += n - 1 - pos
		}
	}

	maxScore := 0
	for _, score := range scores {
		if score > maxScore {
			maxScore = score
		}
	}

	winners := []int{}
	for i, score := range scores {
		if score == maxScore {
			winners = append(winners, i)
		}
	}
	return winners
}

// Check Condorcet winner
func isCondorcetWinner(c int, prefs [][]int, n, k int) bool {
	for other := 0; other < n; other++ {
		if other == c {
			continue
		}

		wins := 0
		for _, ranking := range prefs {
			cPos := indexOf(ranking, c)
			otherPos := indexOf(ranking, other)
			if cPos < otherPos {
				wins++
			}
		}

		if wins <= k/2 {
			return false
		}
	}
	return true
}

func condorcetWinner(prefs [][]int, n, k int) int {
	for c := 0; c < n; c++ {
		if isCondorcetWinner(c, prefs, n, k) {
			return c
		}
	}
	return -1
}

// Helper function to find index
func indexOf(slice []int, val int) int {
	for i, v := range slice {
		if v == val {
			return i
		}
	}
	return -1
}

func main() {
	reader := bufio.NewReader(os.Stdin)

	var n, k int
	fmt.Print("Введите количество кандидатов (n): ")
	fmt.Scan(&n)
	fmt.Print("Введите количество избирателей (k): ")
	fmt.Scan(&k)
	reader.ReadString('\n') // Clear buffer

	// Read candidates
	candidates := make([]string, n)
	candidateMap := make(map[string]int)
	fmt.Println("Введите имена кандидатов (как есть, без нормализации):")
	for i := 0; i < n; i++ {
		fmt.Printf("Кандидат %d: ", i+1)
		name, _ := reader.ReadString('\n')
		name = strings.TrimSpace(name)
		
		if _, exists := candidateMap[name]; exists {
			fmt.Fprintf(os.Stderr, "Ошибка: Кандидат '%s' уже существует!", name)
			os.Exit(1)
		}
		
		candidates[i] = name
		candidateMap[name] = i
	}

	// Read preferences
	prefs := make([][]int, k)
	fmt.Printf("\nВведите предпочтения (%d имен через пробел):\n", n)
	for i := 0; i < k; i++ {
		for {
			fmt.Printf("Избиратель %d: ", i+1)
			input, _ := reader.ReadString('\n')
			input = strings.TrimSpace(input)
			names := strings.Fields(input)

			if len(names) != n {
				fmt.Printf("Ошибка: Нужно %d кандидатов!\n", n)
				continue
			}

			valid := true
			votes := make([]int, n)
			unique := make(map[int]bool)
			
			for j, name := range names {
				if idx, exists := candidateMap[name]; exists {
					if unique[idx] {
						fmt.Println("Ошибка: Повторы в списке!")
						valid = false
						break
					}
					votes[j] = idx
					unique[idx] = true
				} else {
					fmt.Printf("Ошибка: Неизвестный кандидат '%s'\n", name)
					valid = false
					break
				}
			}

			if valid {
				prefs[i] = votes
				break
			}
		}
	}

	// Calculate results
	borda := bordaWinners(prefs, n)
	condorcet := condorcetWinner(prefs, n, k)

	// Print results
	fmt.Println("\nРезультаты:")
	fmt.Print("Метод Борда: ")
	if len(borda) == 1 {
		fmt.Println(candidates[borda[0]])
	} else {
		fmt.Print("Ничья между: ")
		for i, idx := range borda {
			fmt.Printf("'%s'", candidates[idx])
			if i < len(borda)-1 {
				fmt.Print(", ")
			}
		}
		fmt.Println()
	}

	fmt.Print("Метод Кондорсе: ")
	if condorcet != -1 {
		fmt.Println(candidates[condorcet])
	} else {
		fmt.Println("Нет победителя")
	}

	// Check for different winners
	if condorcet != -1 {
		found := false
		for _, c := range borda {
			if c == condorcet {
				found = true
				break
			}
		}
		if !found {
			fmt.Println("\n⚠️ Методы дали разных победителей!")
		}
	}
}
